tmt/utils/signals.py

```diff
@@ -0,0 +1,149 @@
+_INTERRUPT_MASKED = threading.Event()
+
+#: When set, interrupt was delivered to tmt, and tmt should react to it.
+_INTERRUPT_PENDING = threading.Event()
+
+
+def _quit_tmt(logger: tmt.log.Logger, repeated: bool = False) -> NoReturn:
+    """
+    Send tmt on the path of quitting by raising an exception.
+    """
+
+    if repeated:
+        logger.warning(
+            textwrap.dedent(
+                """
+              Repeated interruption requested.
+
+              tmt will now cancel its work in progress and quit as soon as
+              possible. Wait for it to finish, please.
+              """
+            ).strip()
+        )
+
+    else:
+        logger.warning(
+            textwrap.dedent(
+                """
+              Interrupting tmt operation as requested.
+
+              tmt will now cancel its work in progress and quit as soon as
+              possible. Wait for it to finish, please.
+
+              Interrupt tmt again for faster termination but be aware that
+              it may result in resource leaks as various cleanup tasks will
+              not finish.
+              """
+            ).strip()
+        )
+
+    raise KeyboardInterrupt
+
+
+def _interrupt_handler(signum: int, frame: Optional[FrameType]) -> None:
+    """
+    A signal handler for signals that interrupt tmt, ``SIGINT`` and ``SIGTERM`.
+
+    :param signum: delivered signal.
+    :param frame: stack frame active when the signal was received.
+    """
+
+    logger = tmt.log.Logger.get_bootstrap_logger()
+
+    logger.warning(f'Interrupt requested via {signal.Signals(signum).name} signal.')
+
+    with _INTERRUPT_LOCK:
+        repeated = _INTERRUPT_PENDING.is_set()
+
+        _INTERRUPT_PENDING.set()
+
+        if _INTERRUPT_MASKED.is_set():
+            logger.warning('Interrupt is masked, postponing the reaction.')
+
+            return
+
+        _quit_tmt(logger, repeated=repeated)
+
+
+def install_handlers() -> None:
+    """Install tmt's signal handlers"""
+
+    signal.signal(signal.SIGINT, _interrupt_handler)
+    signal.signal(signal.SIGTERM, _interrupt_handler)
+
+
+class PreventSignals(contextlib.AbstractContextManager['PreventSignals']):
+    """
+    For the duration of this context manager, interrupt signals are postponed.
+
+    If, while the context was active, signals were delivered,
+    :py:exc:`KeyboardInterrupt` exception would be raised when leaving
+    the context.
+    """
+
+    logger: tmt.log.Logger
+
+    def __init__(self, logger: tmt.log.Logger) -> None:
+        self.logger = logger
+
+    def __enter__(self) -> 'PreventSignals':
+        with _INTERRUPT_LOCK:
+            _INTERRUPT_MASKED.set()
+
+        return self
+
+    def __exit__(self, *args: object) -> None:
+        with _INTERRUPT_LOCK:
+            _INTERRUPT_MASKED.clear()
+
+            if not _INTERRUPT_PENDING.is_set():
+                self.logger.debug('Interrupt not detected, leaving safe block.')
```

Comment:
Perhaps use `level=2` for these debug lines?

Reply:
Addressed in 677e4f05